src/main/java/com/google/firebase/internal/RetryConfig.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.util.BackOff;
+import com.google.api.client.util.ExponentialBackOff;
+import com.google.api.client.util.Sleeper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Configures when and how HTTP requests should be retried.
+ */
+public final class RetryConfig {
+
+  private static final int INITIAL_INTERVAL_MILLIS = 500;
+
+  private final List<Integer> retryStatusCodes;
+  private final int maxRetries;
+  private final Sleeper sleeper;
+  private final ExponentialBackOff.Builder backOffBuilder;
+
+  private RetryConfig(Builder builder) {
+    if (builder.retryStatusCodes != null) {
+      this.retryStatusCodes = ImmutableList.copyOf(builder.retryStatusCodes);
+    } else {
+      this.retryStatusCodes = ImmutableList.of();
+    }
+
+    checkArgument(builder.maxRetries >= 0, "maxRetries must not be negative");
+    this.maxRetries = builder.maxRetries;
+    this.sleeper = checkNotNull(builder.sleeper);
+    this.backOffBuilder = new ExponentialBackOff.Builder()
+        .setInitialIntervalMillis(INITIAL_INTERVAL_MILLIS)
+        .setMaxIntervalMillis(builder.maxIntervalMillis)
```

Comment:
It's not immediately obvious to me why some of these are configureable (max interval, multiplier), while others are not. Have you considered making them all configurable? If that is a direction that we could move to, then we might be able to use `ExponentialBackOff` directly.

Reply:
We have rolled out similar retry mechanisms in Python and Go. I was trying to be consistent with those implementations. The default policy we want to support is:

1. Retry up to N times on selected HTTP status codes
2. Delay subsequent retries with exponential backoff
3. Support the `Retry-After` header when available, as long as it is below a specified maximum

I designed `RetryConfig` to have the minimum set of controls required to configure the above policy. We can add more controls in the future, if necessary. In either case I don't think we want to expose Google API client types to other classes in the SDK. Less exposure to the third-party code, the better.